File: rxjava2/src/test/java/com/example/demo/TestMainVerticle.java
package com.example.demo;



import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import io.vertx.reactivex.core.Vertx;
import io.vertx.reactivex.ext.web.client.WebClient;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.logging.Level;
import java.util.logging.Logger;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(VertxExtension.class)
public class TestMainVerticle {
    private static final Logger LOGGER = Logger.getLogger(TestMainVerticle.class.getName());

    @BeforeEach
    void deploy_verticle(Vertx vertx, VertxTestContext testContext) {
        vertx.deployVerticle(new MainVerticle(), testContext.succeeding(id -> testContext.completeNow()));
    }

    @Test
    void verticle_deployed(Vertx vertx, VertxTestContext testContext) throws Throwable {
        testContext.completeNow();
    }

    @Test
    void testGetAll(Vertx vertx, VertxTestContext testContext) throws Throwable {
        WebClient client = WebClient.create(vertx);
        client.get(8888, "localhost", "/posts")
            .rxSend()
            .subscribe(
                response -> {
                    assertThat(response.statusCode()).isEqualTo(200);
                    assertThat(response.bodyAsJsonArray().size()).isEqualTo(2);

                    testContext.completeNow();
                }
            );
    }

    @Test
    void testPostValidation(Vertx vertx, VertxTestContext testContext) throws Throwable {
        WebClient client = WebClient.create(vertx);
        client.post(8888, "localhost", "/posts")
            .rxSendJson(PostForm.of("test", ""))
            .subscribe(
                response -> {
                    var statusCode = response.statusCode();
                    var body = response.bodyAsString();
                    LOGGER.log(Level.INFO, "status code: {0}, body: {1}", new Object[] {
                        statusCode, body
                    });
                    assertThat(statusCode).isEqualTo(400);
                    assertThat(body).isEqualTo("validation failed.");
                    testContext.completeNow();
                }
            );
    }

}

File: kotlin-co/src/test/kotlin/com/example/demo/TestMainVerticle.kt
package com.example.demo

import com.fasterxml.jackson.core.type.TypeReference
import io.kotest.matchers.equals.shouldBeEqual
import io.vertx.core.Vertx
import io.vertx.core.json.jackson.DatabindCodec
import io.vertx.ext.web.client.WebClient
import io.vertx.ext.web.client.WebClientOptions
import io.vertx.kotlin.coroutines.coAwait
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger
import kotlin.time.Duration.Companion.milliseconds

// see: https://github.com/vietj/kotlin-conf-inter-reactive/blob/master/src/test/kotlin/com/julienviet/movierating/MovieRatingTest.kt
class TestMainVerticle {
    companion object {
        private val LOGGER = Logger.getLogger(TestMainVerticle::class.java.name)
    }

    lateinit var vertx: Vertx
    lateinit var client: WebClient

    @BeforeEach
    fun setUp() = runTest {
        vertx = Vertx.vertx()
        client = WebClient.create(vertx, WebClientOptions().setDefaultPort(8888))

        vertx.deployVerticle(MainVerticle())
            .onComplete {
                LOGGER.log(Level.INFO, "Deployed MainVerticle: $it")
            }
            .coAwait()
    }

    @AfterEach
    fun tearDown() {
        vertx.close()
    }

    @Test
    fun `get all posts`() = runTest {
        val response = client.get("/posts").send().coAwait()
        response.statusCode() shouldBeEqual 200
        LOGGER.info("response :${response.bodyAsString()}")
        val body = response.bodyAsString()
        val posts: List<Post> = DatabindCodec.mapper().readValue(body, object : TypeReference<List<Post>>() {})
        LOGGER.info("posts: ${posts}")
        //posts.size() shouldBeEqual 2
    }

    @Test
    fun `get post by none-existing id`() = runTest(timeout = 500.milliseconds) {
        val id = UUID.randomUUID()
        val response = client.get("/posts/$id").send().coAwait()
        response.statusCode() shouldBeEqual 404
    }
}
